### DISCUSSION TITLE
login decorator for JWT token verification

### DISCUSSION THREAD

**Discussion by Pk13055** [fastapi/fastapi#7636] [Questions] [answered]
### First check

* [X] I used the GitHub search to find a similar issue and didn't find it.
* [X] I searched the FastAPI documentation, with the integrated search.
* [X] I already searched in Google "How to X in FastAPI" and didn't find any information.

I would first off like to thank you, @tiangolo for making an excellent ~alternative to~ replacement of Flask, with amazing features like openapi compat and typing support. (not to mention blazing fast and using uvicorn)

### Description

So, I have authentication being done by [CAS](https://github.com/python-cas/) and I am able to successfully retrieve user info and generate a jwt. However, I am unsure of two things:

1. How can I protect certain backend routes with a decorator? 
```python
def login_required(f):
    @wraps(f)
    def wrap(*args, **kwargs):
        if 'logged_in' in session:
            return f(*args, **kwargs)
        else:
            flash("You need to login first")
            return redirect(url_for('login_page'))
    return wrap
```
(_this question can also be framed as how can I essentially scan each request header for a `Bearer` token when made to a protected API_)

2. How can I send the token to the frontend? 
Now this question seems kind of weird but the issue is my flow currently, viz, frontend makes a `GET` request to `/api/login` -> redirects to external CAS login -> redirects back to `/api/login` -> ??

What I'm thinking is I should basically use something like `axios` to send the `GET` request and then capture the response at the end of the entire cycle, but do let me know if there's a better way to do it.

PS - I think the answer to both has something to do with [this](https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-in-path-operation-decorators/) part of the documentation, but I wasn't able to construct it for my use case.

### Additional context

I am running my backend, FastAPI server using uvicorn inside an alpine docker container served over nginx reverse-proxy. My frontend, is [svelte](https://svelte.dev/) (also reverse-proxy docker nginx)

**Comment by Pk13055** [comment, marked as answer]
This can be done using either `dependencies=[Depends(verify_token)]` wherein verify_token would have `authorization: Header(“Authorization”)` as an arg

**Comment by tiangolo** [comment]
Thanks for reporting back and closing the issue @Pk13055 :+1:

**Comment by flaviorangel** [comment]
> This can be done using either `dependencies=[Depends(verify_token)]` wherein verify_token would have `authorization: Header(“Authorization”)` as an arg

Did you mean `authorization = Header(“Authorization”)`? I tried this solution, using the code below for verify_token function:

```Ruby
from fastapi import HTTPException, Header, Depends

async def verify_token(token: str = Header("Authorization")):
    print("THIS IS TOKEN:", token)
    print(type(token))
    if not token_is_valid(token):
        raise HTTPException(status_code=401, detail="Unauthorized token.")

@router.get("/aurl", status_code=200, tags=["atag"],
            dependencies=[Depends(verify_token)])
async def get_aurl(request: Request):
    # Code...
```

I'm using Postman to send a request with an authorization token, but I can't access it through the fast Header function. Terminal shows:
THIS IS TOKEN: Authorization
<class 'str'>

It seems I get as return whatever I type as argument in Header. Shouldn't it access the token value in the request header? On the other hand, I'm able to access the token using:

```Ruby
@router.get("/aurl", status_code=200, tags=["atag"])
async def get_aurl(request: Request):
    headers = request.headers
    token = headers['Authorization']
```

But then I can't pass the request as argument in `verify_token`

**Comment by flaviorangel** [comment]
Nevermind, I got it! I needed to include `@app.get("/items/")` before verify token. Also, the parameter name should match the one in header:

```Ruby

from fastapi import HTTPException, Header, FastAPI

@app.get("/items/")
async def verify_token(authorization: str = Header("Authorization")):
    if not token_is_valid(authorization):
        raise HTTPException(status_code=401, detail="Unauthorized token.")
```

**Comment by lucasvazq** [comment]
To verify the JWT in the header I do the following:

```python
from jose.jwt import decode
from jose.exceptions import JWTError


async def verify_jwt(authorization: str = Header()):
    jwt_token = authorization.split(" ", 1)[1]
    try:
        decode(jwt_token, "JWT_SECRET_KEY", algorithms="HS256")
    except JWTError:
        raise HTTPException(status_code=401, detail="Unauthorized token")


@router.get("/items/", dependencies=[Depends(verify_jwt)])
async def read_items():
    return [{"item": "Foo"}, {"item": "Bar"}]
```

This forces the user to send their request with the following header: `Authorization: Bearer <TOKEN>`
